sycl/include/CL/sycl/bit_cast.hpp

```diff
@@ -20,6 +22,10 @@ namespace sycl {
 // forward decl
 namespace detail {
 inline void memcpy(void *Dst, const void *Src, std::size_t Size);
+
+template <typename Type> struct type_helper { using T = Type; };
+
+template <> struct type_helper<sycl::half> { using T = uint16_t; };
```

Comment:
half is special type and can lower to different types on host and device. Something similar to https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/detail/generic_type_traits.hpp#L398-L399 should be used.